internal/k8s/replicaset/replicaset.go

```diff
@@ -0,0 +1,142 @@
+//
+// Copyright (C) 2019-2020 Vdaas.org Vald team ( kpango, rinx, kmrmt )
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package replicaset
+
+import (
+	"context"
+	"reflect"
+	"strings"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/log"
+)
+
+// ReplicaSetWatcher is a type alias for k8s resource controller.
+type ReplicaSetWatcher k8s.ResourceController
+
+type reconciler struct {
+	ctx         context.Context
+	mgr         manager.Manager
+	name        string
+	namespace   string
+	onError     func(err error)
+	onReconcile func(rs map[string][]ReplicaSet)
+}
+
+// ReplicaSet is a type alias for the k8s replica set definition.
+type ReplicaSet = appsv1.ReplicaSet
+
+// New returns the ReplicaSetWatcher that implements reconciliation loop, or any error occurred.
+func New(opts ...Option) (ReplicaSetWatcher, error) {
+	r := new(reconciler)
+
+	for _, opt := range opts {
+		if err := opt(r); err != nil {
+			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
+		}
+	}
+
+	return r, nil
+}
+
+// Reconcile implements k8s reconciliation loop to retrieve the ReplicaSet information from k8s.
+func (r *reconciler) Reconcile(req reconcile.Request) (res reconcile.Result, err error) {
+	rsl := new(appsv1.ReplicaSetList)
+
+	err = r.mgr.GetClient().List(r.ctx, rsl)
+	if err != nil {
+		if r.onError != nil {
+			r.onError(err)
+		}
+		res = reconcile.Result{
+			Requeue:      true,
+			RequeueAfter: time.Millisecond * 100,
+		}
+		if k8serrors.IsNotFound(err) {
+			log.Error("not found", err)
```

Comment:
it may be better to use log.xxxf function.
```suggestion
			log.Errorf("not found: %s", err)
```

Is it really ERROR level log? I think it is not so serious situation, just WARN level or lower one.
https://github.com/vdaas/vald/blob/master/docs/contributing/coding-style.md#logging

Reply:
I think it's better to define new error variable in errors package eg. `errors.ErrK8SResourceNotFound(err)`